// @appium/base-driver/build/test/helpers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createAppiumURL = exports.TEST_HOST = exports.METHODS = void 0;
exports.getTestPort = getTestPort;

require("source-map-support/register");

var _getPort = _interopRequireDefault(require("get-port"));

var _lodash = require("lodash");

const TEST_HOST = '127.0.0.1';
exports.TEST_HOST = TEST_HOST;
const METHODS = {
  POST: 'POST',
  DELETE: 'DELETE'
};
exports.METHODS = METHODS;
let testPort;

async function getTestPort(force = false) {
  if (force || !testPort) {
    let port = await (0, _getPort.default)();

    if (!testPort) {
      testPort = port;
    }

    return port;
  }

  return testPort;
}

const createAppiumURL = (0, _lodash.curry)((address, port, session, pathname) => {
  if (!/^https?:\/\//.test(address)) {
    address = `http://${address}`;
  }

  let path = session ? `session/${session}` : '';

  if (pathname) {
    path = `${path}/${pathname}`;
  }

  let url = new URL(path, `${address}:${port}`);
  return url.href;
}, 4);
exports.createAppiumURL = createAppiumURL;
