// @appium/test-support/build/lib/sandbox-utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.verifySandbox = verifySandbox;
exports.withSandbox = withSandbox;

require("source-map-support/register");

var _sinon = _interopRequireDefault(require("sinon"));

var _lodash = _interopRequireDefault(require("lodash"));

var _bluebird = _interopRequireDefault(require("bluebird"));

let SANDBOX = Symbol();

function withSandbox(config, fn) {
  return () => {
    const S = {
      mocks: {},

      verify() {
        return this.sandbox.verify();
      }

    };
    beforeEach(function beforeEach() {
      S.sandbox = _sinon.default.createSandbox();
      S.sandbox.usingPromise(_bluebird.default);
      S.mocks[SANDBOX] = S.sandbox;

      for (let [key, value] of _lodash.default.toPairs(config.mocks)) {
        S.mocks[key] = S.sandbox.mock(value);
      }
    });
    afterEach(function afterEach() {
      S.sandbox.restore();

      for (let k of _lodash.default.keys(S.mocks)) {
        delete S.mocks[k];
      }

      delete S.mocks[SANDBOX];
    });
    fn(S);
  };
}

function verifySandbox(obj) {
  let sandbox = obj.sandbox ? obj.sandbox : obj[SANDBOX];
  sandbox.verify();
}
